org-cyf-sdc/content/tools/sprints/3/process_spelunking_workshop.md

```diff
@@ -0,0 +1,539 @@
+| 1               | `stdout` (standard output) | Your program can write output on this *file descriptor*.                            |
+| 2               | `stderr` (standard error)  | If things go wrong in your program, you can write errors to this *file descriptor*. |
+
+The *standard input* is used by our tiny program, to wait for the "enter" key.
+And the message `Press enter to stop the program...` is written to standard output.
+
+These *file descriptors* (0, 1, and 2) always exist, they're created at the start of the program.
+This is also why the first opened file ends up on *file descriptor* `3`: it's the first available number that can be used.
+
+### Reg, Chr, Dir?
+
+The `TYPE` column tells us what kind of file is opened.
+`REG` means something that is a regular file.
+Because we created a regular file for our program, we can see that `exercise_2_file.txt` is opened as a `REG` file.
+
+All other types are special kinds of files.
+In particular, `DIR` means directory.
+(Yes, you can open directories. The directory mentioned under *FD* `cwd`, is the Current Working Directory, where you program is running.)
+
+And `CHR` means character special device.
+These are usually things like the console.
+Which is why you see *standard input*, *standard output*, and *standard error* having those types of file open.
+
+There are many many more types, but you won't see those quite as often.
+You can read them all in the `lsof` man-page: `man lsof`.
+(Most of them I don't know, and have never seen in the wild.)
+
+{{<note title="Clean up">}}
+Now that you know how to find a *File Descriptor*, we no longer need the program that we started on the first terminal.
+You can hit "enter" to get it to stop.
+{{</note>}}
+
+### What You've Learned
+
+Processes can open files.
+They use *file descriptors* to refer to the files.
+And we can inspect which files are opened by a process, using the `lsof` command.
+
+
+
+## System Calls
+
+So far, we've seen files getting opened.
+And we've seen we can use `lsof` to see those files being opened.
+
+The reason `lsof` can see it, is because the process asks the kernel.
+The kernel was asked for a file, thus it knows the file is opened, and can tell you about it.
+
+This "asking the kernel" is done by use of *system calls*.
+
+### Running strace
+
+We can look at *system calls* while they happen.
+`strace` is short for *system call trace*.
+
+{{<note title="Preparation">}}
+Create a new file `exercise_3_file.txt`.
+This time, we want to put something in it.
+You can write what you like, but if you're short on inspiration, write this:
+```
+Agnes, who is a cat, is sleeping on the sofa.
+```
+
+Create a file `exercise_3.py` with this code.
+```
+file_object = open("exercise_3_file.txt", "r")
+```
+This time, the program doesn't wait for you to press enter.
+You can test if you've set everything up correctly, by running `python3 exercise_3.py`.
+It should write the text you wrote in `exercise_3_file.txt`.
```

Comment:
I'm not sure about this check - the program doesn't currently output anything, so it's hard to check. Maybe we should rephrase this to something like "We expect the program to successfully exit with no output."

We could even use this as an excuse to introduce `$?` which they probably haven't encountered before?

Reply:
Done.